pkg/storaged/dialogx.jsx

```diff
@@ -462,3 +464,128 @@ export const Skip = (className, options) => {
+        document.addEventListener("mouseup", stop_dragging);
+        drag(event);
+    }
+
+    if (fraction < 0) fraction = 0;
+    if (fraction > 1) fraction = 1;
+
+    return (
+        <div className="slider" onMouseDown={start_dragging}>
+            <div className="slider-bar" style={{ width: fraction * 100 + "%" }}>
+                <div className="slider-thumb" />
+            </div>
+        </div>
+    );
+};
+
+class SizeSliderElement extends React.Component {
+    constructor(props) {
+        super();
+        this.units = cockpit.get_byte_units(props.value || props.max);
+        this.units.forEach(u => { if (u.selected) this.state = { unit: u.factor }; });
+    }
+
+    render() {
+        let { val, max, onChange } = this.props;
+        let { unit } = this.state;
+
+        const change_slider = (f) => {
+            onChange(Math.round(f * max));
+        };
+
+        const change_text = (event) => {
+            if (event.type == "change") {
+                let val = +event.target.value * unit;
+                if (event.target.value === "" || isNaN(val)) {
+                    /* If there something else than a number in the
+                       input element, we use that as the value
+                       directly so that it sticks around.  It will be
+                       rejected by the validate function below.
+                     */
+                    onChange(event.target.value);
+                } else {
+                    onChange(Math.round(val));
+                }
+            }
+        };
+
+        const change_unit = (u) => {
+            this.setState({ unit: +u });
+        };
+
+        return (
+            <div className="size-sliderx">
+                <StatelessSlider fraction={val / max} onChange={change_slider} />
+                <input className="size-text form-control"
+                       value={ val === "" || isNaN(val) ? val : cockpit.format_number(val / unit) }
+                       onChange={change_text} />
+                <StatelessSelect extraClass="size-unit" selected={unit} onChange={change_unit}>
+                    { this.units.map(u => <SelectEntry data={u.factor} key={u.name}>{u.name}</SelectEntry>) }
+                </StatelessSelect>
+            </div>
+        );
+    }
+}
+
+export const SizeSlider = (tag, title, options) => {
+    let validate = (val, vals) => {
+        let msg = null;
+
+        if (val === "" || isNaN(val))
+            msg = _("Size must be a number");
+        if (val === 0)
+            msg = _("Size cannot be zero");
+        if (val < 0)
+            msg = _("Size cannot be negative");
+        if (!options.allow_infinite && val > options.max)
+            msg = _("Size is too large");
+        if (options.min !== undefined && val < options.min)
+            msg = cockpit.format(_("Size must be at least $0"), fmt_size(options.min));
+
+        if (options.validate)
+            msg = options.validate(val, vals);
```

Comment:
Shoudl this be an if-else-if ladder? Otherwise multiple conditions could potentially match.

Reply:
It could be.  I don't think there are any overlapping cases, but we could use the ladder to avoid calling the external `validate` function with arbitrary values.  I'll change this.